### ISSUE TITLE
Add MPI_Timer requests

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#277]
**Originally by _ftillier_ on 2011-06-03 01:14:39 -0500**

---

MPI provides two different semantics for checking for completion of non-blocking requests: instantaneous (e.g. MPI_Test) and blocking (e.g. MPI_Wait).  There is no way to make strong progress for a limited amout of time, the missing middle between the instantaneous and the blocking endpoints.
## Proposed Solution

This proposal intorduces timer requests, a new type of non-blocking request that completes locally after a given due time.

A timer request is created by the MPI_Timer_create function, taking as input a timeout whereby the request should complete when MPI_Wtime >= timeout.

Such a timer request can be used in any Test/Wait routine along with other non-blocking requests.

When a timer request completes, it is freed as with any other request.  Calling MPI_Request_free on a timer request cancels and frees the timer in a single operation, without need for waiting for cancelation.

A pending timer request can have its timeout reset by calling MPI_Timer_reset.  This removes the need to create new timers if a non-timer request satisfied a wait by allowing existing timers to be reset and reused.
## Alternate Designs

One might suggest introducing new wait functions in the API that take an additional timeout parameter.  Such a solution would need not only a timeout parameter, but also a flag to indicate whether the wait timed out.  Instead of a flag one could define an MPI_TIMEOUT recoverable error code, however a timeout may not necessarilly be an error, so reporting such via a non-success error code is unnatural.

Another possibility would be for users to use the existing generalized request interface, manage their own timers, and complete the generalized request themselves at the appropriate time.  However, this does not work with MPI implementations providing MPI_THREAD_SERIALIZED support since a thread would already have entered the MPI library to perform the wait, so another thread could not enter to mark a request as complete.  Any other mechanism of interrupting the MPI library, such as using signals, is non-portable.  Likewise, requiring users to manage their own timers makes their codes non-portable, as they must deal with a timer support routines for each target platform themselves, rather than letting the MPI implementation abstract the differences.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2011-06-03 07:47:43 -0500**

---

**Attachment added:** [`timer-requests.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/277/timer-requests.pdf) (357.6 KiB)
Proposal slides for MPI timer requests

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-06-09 13:17:15 -0500**

---

**Attachment added:** [`ticket-277-2-pages.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/277/ticket-277-2-pages.pdf) (162.7 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-06-09 13:29:15 -0500**

---

**Attachment added:** [`ticket-277.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/277/ticket-277.pdf) (2283.4 KiB)
Full standard document with ticket 277 changes.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2011-07-19 05:55:59 -0500**

---

Got a lot of good feedback from the Forum after the reading today.
- change "due time" to timeout throughout the text.  The Forum felt that "due time" implies absolute times.
- change:
  For example, a call to MPI_TIMER_CREATE with due_time # 5 at time AMPI_WTIME will result in the timer request completing when MPI_WTIME returns A + 5 or greater.
- The Forum felt this was awkward (e.g., they didn't like the "A=MPI_WTIME" syntax because it makes MPI_WTIME look like a constant).  They wanted this version instead:
  For example, a call to MPI_TIMER_CREATE with due_time = 5.0 will result in the timer request completing no sooner than 5.0 seconds after the call to MPI_TIMER_CREATE.
- need to add something about how timer requests will "eventually complete" -- avoid the "Pavan" scenario where an implementation can ignore the timer completely (because we say that they can complete at any time in the future).  Given that, Torsten made the point that he could write a legal code that will never complete.  So we need _something_ -- even if it's weak, a la MPI_SEND -- that timer requests must eventually complete.
- Rolf makes the point that all the text about MPI_CANCEL is about communciation.  We should specify what it means in terms of timers.
- change: "It saves the step of canceling an active timer and creating a new one." to say something about canceling and _completeing or freeing_ the old request.
- change: "This function is provided as a convenience for cases where applications need to reset a timer **request** that has not yet completed."
- feedback from rolf: we don't define what goes into the status after successful test or wait.
- feedback from aurilean: might be useful to specifically state that RSTART_TIMER can _only_ be used with timer requests; passing in any other request will... return a specific error code? be undefined? be erroneous?
- perhaps we should have an example that shows both MPI_WAITANY and MPI_WAITALL...?
- brian doesn't like this because you now have to scan through all the reqs in array TEST/WAIT looking for timers.  Answer: you already have to for generalized requests.
- significant resistance from torsten:
  - he can't think of a good use case.
  - he doesn't think timers have anything to do with message passing.
  - he points out that the beginning of the wtick/wtime section specifically says that timers in MPI are not about message passing; they are only for debugging programs.
  - if we opened MPI up to more than just HPC, he feels that this would be more applicable.  But the Forum specifically restricted itself to an HPC context at the beginning of MPI3.
  - he did not think the use case of polling MPI and something else was sufficient -- his counter was looping over MPI_TEST for a specific amount of time.
  - torsten felt _very_ strongly about this.
- note, too, that the beginning of the wtime/wtick section says that MPI has timers, but that does not refer to our timer requests.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2011-07-19 05:58:18 -0500**

---

One more note from Auri -- a use case example that they have:

  I think I do have an example where timed requests can be beneficial, but I am not certain the behavior cannot be emulated otherwise. On XT5 machines, the scheduling time quantum is gigantic, and a thread that does a yield or a nanosleep gets descheduled quickly. We have this DAG engine that executes multithread applications and uses an extra thread (MPI_Funneled) to do all communications. That thread must alternate between probing the MPI, probing the command queue from the compute thread, and must yield when it has nothing to do so that it does not disturb computation with busy loops. The current implementation uses nanosleep for that later purpose, which works fine most of the time, but is uterly broken on the XT5 hardware, because latency becomes so huge that the entire application stalls to a crawl. Somehow, having the MPI  do the "nanosleep" by hardware control might (or might not) help.

**Comment by mpiforumbot** [comment]
**Originally by _jjhursey_ on 2012-01-31 09:13:23 -0600**

---

Another possible use case was mentioned recently on the Fault Tolerance working group listserv:
- http://lists.mpi-forum.org/mpi3-ft/2012/01/0976.php

Adding the author to this ticket since they might be interested in its progress.

**Comment by mpiforumbot** [comment]
**Originally by @wbland on 2013-05-13 11:14:14 -0500**

---

This ticket hasn't been touched in quite a while and it appears that all who were involved in it have moved on. There's a discussion on the FT working group to withdraw the ticket at the end of June '13 if no one steps up to defend/revive this work.

[http://lists.mpi-forum.org/mpi3-ft/2013/05/1251.php]

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2013-05-13 11:33:54 -0500**

---

I wouldn't say that those who were involved (Jeff Squyres and myself) have moved on.  This ticket is on hold, due to folks pushing back on the basis that generalized requests were intended to enable this kind of functionality.  The issue at hand is that generalized requests are only useful/usable in a MPI_THREAD_MULTIPLE environment, and we're looking at whether that can be fixed, rather than adding this new functionality.  If it can't, we'll bring this back up for (more) discussion.

**Comment by mpiforumbot** [comment]
**Originally by @wbland on 2013-05-13 12:35:14 -0500**

---

Noted. We won't drop the ticket then.

Replying to ftillier:

> I wouldn't say that those who were involved (Jeff Squyres and myself) have moved on.  This ticket is on hold, due to folks pushing back on the basis that generalized requests were intended to enable this kind of functionality.  The issue at hand is that generalized requests are only useful/usable in a MPI_THREAD_MULTIPLE environment, and we're looking at whether that can be fixed, rather than adding this new functionality.  If it can't, we'll bring this back up for (more) discussion.
